modules/core/src/system.cpp

```diff
@@ -936,241 +936,274 @@ bool Mutex::trylock() { return impl->trylock(); }
-#endif
-    static DWORD tlsKey = TLS_OUT_OF_INDEXES;
+    std::vector<void*> slots; // Data array for a thread
+    size_t idx;               // Thread index in TLS storage. This is not OS thread ID!
+};
 
-    static void deleteThreadData()
+// Main TLS storage class
+class TlsStorage
+{
+public:
+    TlsStorage()
     {
-        if(tlsKey != TLS_OUT_OF_INDEXES)
-        {
-            delete (TLSStorage*)TlsGetValue(tlsKey);
-            TlsSetValue(tlsKey, NULL);
-        }
+        tlsSlots = 0;
+        threads.reserve(32);
     }
-
-    inline TLSStorage* TLSStorage::get()
+    ~TlsStorage()
     {
-        if (tlsKey == TLS_OUT_OF_INDEXES)
+        for(size_t i = 0; i < threads.size(); i++)
         {
-            tlsKey = TlsAlloc();
-            CV_Assert(tlsKey != TLS_OUT_OF_INDEXES);
-        }
-        TLSStorage* d = (TLSStorage*)TlsGetValue(tlsKey);
-        if (!d)
-        {
-            d = new TLSStorage;
-            TlsSetValue(tlsKey, d);
+            if(threads[i])
+            {
+                /* Current architecture doesn't allow proper global objects relase, so this check can cause crashes
+
+                // Check if all slots were properly cleared
+                for(size_t j = 0; j < threads[i]->slots.size(); j++)
+                {
+                    CV_Assert(threads[i]->slots[j] == 0);
+                }
+                */
+                delete threads[i];
+            }
         }
-        return d;
+        threads.clear();
     }
-#endif //WINRT
-
-#if defined CVAPI_EXPORTS && defined WIN32 && !defined WINCE
-#ifdef WINRT
-    #pragma warning(disable:4447) // Disable warning 'main' signature found without threading model
-#endif
 
-extern "C"
-BOOL WINAPI DllMain(HINSTANCE, DWORD fdwReason, LPVOID lpReserved)
-{
-    if (fdwReason == DLL_THREAD_DETACH || fdwReason == DLL_PROCESS_DETACH)
+    void releaseThread()
     {
-        if (lpReserved != NULL) // called after ExitProcess() call
+        AutoLock guard(mtxGlobalAccess);
+        ThreadData *pTD = (ThreadData*)tls.GetData();
+        for(size_t i = 0; i < threads.size(); i++)
         {
-            cv::__termination = true;
-        }
-        else
-        {
-            // Not allowed to free resources if lpReserved is non-null
-            // http://msdn.microsoft.com/en-us/library/windows/desktop/ms682583.aspx
-            cv::deleteThreadAllocData();
-            cv::deleteThreadData();
+            if(pTD == threads[i])
+            {
+                threads[i] = 0;
+                break;
+            }
         }
+        tls.SetData(0);
+        delete pTD;
     }
-    return TRUE;
-}
-#endif
-
-#else
-    static pthread_key_t tlsKey = 0;
-    static pthread_once_t tlsKeyOnce = PTHREAD_ONCE_INIT;
 
-    static void deleteTLSStorage(void* data)
+    // Reserve TLS storage index
+    size_t reserveSlot()
     {
-        delete (TLSStorage*)data;
+        AutoLock guard(mtxGlobalAccess);
+        tlsSlots++;
+        return (tlsSlots-1);
```

Comment:
The ultimate solution would be to use shared vector of keys, which we could updated. But this would require additional sync points for get operation, which can bring significant slowdown.